### PR TITLE
break(hooks): hooks and events system has been merged

### LINKED ISSUE #14117: [Feature]: Combine events and hooks into a single system
### Describe the feature

Internally there is no real difference between hooks and events (other than the intended usage). Combining these will reduce some of our API code and will also make things easier to understand.

During this exercise we could also move to https://symfony.com/doc/current/components/event_dispatcher.html for the registration and triggers.

### Is your feature request related to a problem?

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jdalsem**
fixes #14117

- [x] Update docs
- [x] Add the event dispatcher

### CODE REVIEW

File: engine/classes/Elgg/Router.php [hunk 1]
```diff
@@ -86,7 +86,7 @@ public function __construct(
 	/**
 	 * Routes the request to a registered page handler
 	 *
-	 * This function triggers a plugin hook `'route', $identifier` so that plugins can
+	 * This function triggers a  `'route', $identifier` event so that plugins can
```
Comment (jeabakker) [inline engine/classes/Elgg/Router.php:89]: double space

**Review by jeabakker — COMMENTED** [review]
review part 1

File: engine/tests/phpunit/integration/Elgg/Notifications/NotificationsServiceIntegrationTestCase.php [hunk 2]
```diff
@@ -1028,19 +1017,19 @@ public function testCanUseHooksBeforeAndAfterInstantNotificationsQueue() {
...
-			$this->assertEquals($event->toObject(), $hook->getParam('event')->toObject());
-			$this->assertEquals($subscribers, $hook->getParam('subscriptions'));
+			$this->assertEquals($event->toObject(), $elgg_event->getParam('event')->toObject());
+			$this->assertEquals($subscribers, $elgg_event->getParam('subscriptions'));
 
 			return false;
 		});
 
-		$this->hooks->registerHandler('send:after', 'notifications', function (\Elgg\Hook $hook) use (&$after_call_count, $event, $subscribers) {
+			$this->events->registerHandler('send:after', 'notifications', function (\Elgg\Event $elgg_event) use (&$after_call_count, $event, $subscribers) {
```
Comment (jeabakker) [inline engine/tests/phpunit/integration/Elgg/Notifications/NotificationsServiceIntegrationTestCase.php:1028]: where did the tab come from

File: engine/tests/phpunit/unit/Elgg/HooksRegistrationService/MethodMatcherUnitTest.php [hunk 3]
```diff
@@ -1,6 +1,6 @@
 <?php
 
-namespace Elgg\HooksRegistrationService;
+namespace Elgg\EventsService;
```
Comment (jeabakker) [inline engine/tests/phpunit/unit/Elgg/HooksRegistrationService/MethodMatcherUnitTest.php:3]: If you change the namespace you should move the file to the correct folder

**Review by jeabakker — COMMENTED** [review]
review part 2

**Review by jeabakker — CHANGES_REQUESTED** [review]
see previous reviews

File: engine/classes/ElggPlugin.php [hunk 4]
```diff
@@ -1067,12 +1067,18 @@ protected function deactivateEntities(): void {
 	/**
 	 * Registers the plugin's hooks provided in the plugin config file
 	 *
+	 * @note using hooks in the static config is deprecated
+	 *
```
[the remark this answers is not in the document]
Comment (jdalsem) [reply engine/classes/ElggPlugin.php:1071]: we still use this internally... therefore it is not deprecated

File: docs/design/events.rst [hunk 5]
```diff
@@ -64,12 +44,10 @@ around the state of the system while in flux. E.g. Is the user
...
 
 Before Events have names ending in ":before" and are triggered before
-something happens. Like traditional events, handlers can cancel the
-event by returning ``false``.
+something happens. Handlers can cancel the event by returning ``false``.
 
 After Events, with names ending in ":after", are triggered after
-something happens. Unlike traditional events, handlers *cannot* cancel
-these events; all handlers will always be called.
+something happens. Handlers *cannot* cancel these events; all handlers will always be called.
```
Comment (jeabakker) [inline docs/design/events.rst:50]: something happened

File: docs/appendix/upgrade-notes/3.0-to-3.1.rst [hunk 6]
```diff
@@ -51,7 +51,7 @@ The legacy style hook and event callback arguments are deprecated. You should sw
 	}
 	
 	// new style hook callback
-	function hook_callback(\Elgg\Hook $hook) {
+	function hook_callback(\Elgg\Event $event) {
```
Comment (jeabakker) [inline docs/appendix/upgrade-notes/3.0-to-3.1.rst:54]: not sure if this should be changed in the docs

File: docs/design/events.rst [hunk 7]
```diff
@@ -96,9 +74,6 @@ Elgg event handlers are callables:
 In ``event_handler``, the ``Event`` object has various methods for getting the name, object type,
 and object of the event. See the ``Elgg\Event`` class for details.
 
-If a handler returns ``false``, the event is cancelled, preventing execution of the
-other handlers. All other return values are ignored.
```
Comment (jeabakker) [inline docs/design/events.rst:100]: shouldn't this be added the the 'before' event docs

File: engine/tests/phpunit/integration/Elgg/Notifications/NotificationsServiceIntegrationTestCase.php [hunk 8]
```diff
@@ -486,19 +475,19 @@ public function testCanUseHooksBeforeAndAfterSubscriptionNotificationsQueue() {
...
-			$this->assertEquals(unserialize(serialize($event)), $hook->getParam('event'));
-			$this->assertEquals($subscribers, $hook->getParam('subscriptions'));
+			$this->assertEquals(unserialize(serialize($event)), $elgg_event->getParam('event'));
+			$this->assertEquals($subscribers, $elgg_event->getParam('subscriptions'));
 
 			return false;
 		});
 
-		$this->hooks->registerHandler('send:after', 'notifications', function (\Elgg\Hook $hook) use (&$after_call_count, $event, $subscribers, $object) {
+			$this->events->registerHandler('send:after', 'notifications', function (\Elgg\Event $elgg_event) use (&$after_call_count, $event, $subscribers, $object) {
```
Comment (jeabakker) [inline engine/tests/phpunit/integration/Elgg/Notifications/NotificationsServiceIntegrationTestCase.php:486]: where did the tab come from???

**Review by jeabakker — CHANGES_REQUESTED** [review]

**Review by jeabakker — APPROVED** [review]

**Pull request merged**